File: leetcode209.py
class Solution:
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        ls = len(nums)
        res = ls+1
        start = 0
        end = 0
        val = 0
        while end < ls:
            while val < s:
                val += nums[end]
                end += 1

            while val >= s:

                res = min(res, end - start)
                val -= nums[start]
                start += 1

        return 0 if res == ls+1 else res

if __name__ == "__main__":
    s = Solution()
    print(s.minSubArrayLen(7, [2,3,1,2,4,3]))

File: leetcodec++/leetcode209.cpp
//
// Created by 黄敏辉 on 2019-09-02.
//

class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        int res = 0;
        int start = 0, end = 0;
        int sum = 0;
        while (end < nums.size()){
            sum += nums[end];
            end++;
            if (sum >= s){
                if (res == 0)
                    res = end - start;
                else
                    res = min(res, end - start);
            }
            while (sum - nums[start] >= s){
                sum -= nums[start];
                start++;
                res = min(res, end - start);
            }
        }
        return res;
    }
};